app/jobs/hud_reports/store_artifacts_and_cleanup_job.rb

```diff
@@ -0,0 +1,180 @@
+###
+# Copyright 2016 - 2025 Green River Data Analysis, LLC
+#
+# License detail: https://github.com/greenriver/hmis-warehouse/blob/production/LICENSE.md
+###
+
+# frozen_string_literal: true
+
+# Background job to store HUD report artifacts in S3 and immediately cleanup RDS data
+# This job runs after a report is completed to archive intermediate data and free up RDS space
+module HudReports
+  class StoreArtifactsAndCleanupJob < BaseJob
+    queue_as ENV.fetch('DJ_LONG_QUEUE_NAME', :long_running)
+
+    def perform(report_id)
+      report = HudReports::ReportInstance.find(report_id)
+
+      Rails.logger.info "Processing report #{report_id} for S3 storage and RDS cleanup"
+
+      begin
+        # Step 1: Store artifacts to S3
+        store_artifacts_to_s3(report)
+
+        # Step 2: Update report status (mark as artifacts stored)
+        update_report_status(report)
+
+        # Step 3: Clean up RDS data
+        cleanup_rds_data(report)
+
+        Rails.logger.info "Successfully processed report #{report_id}: stored to S3 and cleaned RDS"
+      end
+    end
+
+    private
+
+    def store_artifacts_to_s3(report)
+      Rails.logger.info "Storing artifacts to S3 for report #{report.id}"
+
+      service = HudReports::S3ArtifactService.new(report)
+      service.store_artifacts!
+
+      Rails.logger.info "Successfully stored artifacts to S3 for report #{report.id}"
+    end
+
+    def cleanup_rds_data(report)
+      Rails.logger.info "Cleaning up RDS data for report #{report.id}"
+
+      # Clean up universe members
+      cleanup_universe_members(report)
+
+      # Clean up report-specific client tables
+      cleanup_report_client_tables(report)
+
+      Rails.logger.info "Successfully cleaned up RDS data for report #{report.id}"
+    end
+
+    def cleanup_universe_members(report)
+      cell_ids = report.report_cells.pluck(:id)
+
+      return unless cell_ids.any?
+
+      deleted_count = HudReports::UniverseMember.where(report_cell_id: cell_ids).delete_all
+      Rails.logger.info "Deleted #{deleted_count} universe members for report #{report.id}"
+    end
+
+    def cleanup_report_client_tables(report)
+      report_version = report.options&.dig('report_version')&.to_sym
+
+      if report_version
+        # Find the generator class from the config using the report name
+        generator_class = generator_class(report.report_name, report_version)
+
+        if generator_class
+          # Get the client classes for this specific generator
+          client_classes = client_classes_for_generator(generator_class)
+
+          client_classes.each do |table_class|
+            if table_class.column_names.include?('report_instance_id')
+              deleted_count = table_class.where(report_instance_id: report.id).delete_all
```

Comment:
This feels brittle, let's force the caller to tell the cleaner-up-er-er what records to clean.